google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClientFactory.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.data.v2;
+
+import com.google.api.core.BetaApi;
+import com.google.api.gax.core.BackgroundResource;
+import com.google.api.gax.core.FixedCredentialsProvider;
+import com.google.api.gax.core.FixedExecutorProvider;
+import com.google.api.gax.rpc.ClientContext;
+import com.google.api.gax.rpc.FixedHeaderProvider;
+import com.google.api.gax.rpc.FixedTransportChannelProvider;
+import com.google.api.gax.rpc.FixedWatchdogProvider;
+import com.google.api.gax.rpc.StubSettings;
+import java.io.IOException;
+import javax.annotation.Nonnull;
+
+/**
+ * A factory to create multiple {@link BigtableDataClient} instances that all share the same channel
+ * pool.
+ *
+ * <p>This allows multiple client instances to share the same gRPC channel pool, which makes client
+ * creation very cheap. The intended use case is for applications that need to access multiple
+ * Bigtable Instances from the same process.
+ *
+ * <p>Example Usage:
+ *
+ * <pre>{@code
+ * BigtableDataSettings defaultSettings = BigtableDataSettings.newBuilder()
+ *   .setProject("my-default-project")
+ *   .setInstance("my-default-instance")
+ *   .build();
+ *
+ * BigtableDataClientFactory clientFactory = BigtableDataClientFactory.create(defaultSettings);
+ *
+ * // Create a new client for "my-default-instance" in "my-default-project";
+ * BigtableDataClient defaultInstanceClient = clientFactory.createDefault();
+ *
+ * // Create a new client for a different application profile
+ * BigtableDataClient otherAppProfileClient = clientFactory.createForAppProfile("other-app-profile");
+ *
+ * // Create a new client for a completely different instance and application profile.
```

Comment:
what happens if a user puts in an already used project/instance/profile?

Reply:
then they'll get the same result as calling createDefault(). Shouldn't be a problem

Reply:
sgtm, minor nit to add a test for that

Reply:
there is nothing special to test for here...it doesnt matter if the project/instance/profile match the default or not